### PR TITLE
Resolves #81

### LINKED ISSUE #81: Create a base class for saving and loading

### PR DESCRIPTION
**Pull request by CinquilCinquil**
#81

### CODE REVIEW

**Comment by CinquilCinquil** [comment]
I saw that there was already a [savable](https://github.com/UFRN-URNAI/urnai-tools/blob/main/urnai/base/savable.py) class on the previous version of URNAI, so i tested it, fixed some bugs and i think its ok to use now.

File: urnai/base/savable.py [hunk 1]
```diff
@@ -0,0 +1,225 @@
+import os
+import pickle
+import tempfile
+from multiprocessing import Process
+
+#import time
+#from urnai.utils.reporter import Reporter as rp
```
Comment (alvarofpp) [inline urnai/base/savable.py]: Remove this.

File: urnai/base/savable.py [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+
+
+class SavableAttr:
+    def __init__(self, value):
+        self.value = value
```
Comment (alvarofpp) [inline urnai/base/savable.py]: I think we can remove this and use a `Dict` instead.

File: urnai/base/savable.py [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+class Savable:
+    """
+    This interface represents the concept of a class that can be saved to disk.
+    The heir class should define a constant or attribute as a default filename
+	to save on disk.
+    """
+
+    def __init__(self, threaded_saving=False):
+        self.threaded_saving = threaded_saving
+        self.pickle_black_list = []
```
Comment (alvarofpp) [inline urnai/base/savable.py]: `blacklist` is a racist term. Use `blocklist` instead.

Comment (alvarofpp) [inline urnai/base/savable.py]: `pickle` restricts the type of file that will be saved. I would use a more general term, as we intend to use several types of files.

File: urnai/base/savable.py [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+		+ 'tensorflow_')
+
+    def get_full_persistance_path(self, persist_path):
+        """This method returns the default persistance path."""
+        return persist_path + os.path.sep + self.get_default_save_stamp()
+
+    def get_full_persistance_pytorch_path(self, persist_path):
+        """This method returns the default persistance pytorch path."""
+        return (persist_path + os.path.sep + self.get_default_save_stamp()
+		+ '.pt')
```
Comment (alvarofpp) [inline urnai/base/savable.py]: I would remove all these functions, with the exception of `get_full_persistance_path`.

**Review by alvarofpp — CHANGES_REQUESTED** [review]
I noticed that `load` methods are in the `Savable` class, I think they should be in another class or we could change the name of `Savable` to something that encompasses these functions.

File: urnai/base/savable.py (hunk 2 above)
Comment (CinquilCinquil) [reply urnai/base/savable.py]: I noticed that this class wasn't being used for anything, so i just removed it, ok?

**Comment by CinquilCinquil** [comment]
For testing, i woud suggest first creating a class, that inherits from Savable, with some attributes. Then instantiate it, giving random values to the attributes, after that save it and load it to see if it still has the same values as before. I belive all methods of the class are used in this procedure, and that should reveal if any of them are not functional.

**Comment by CinquilCinquil** [comment]
Should i make unit tests?

File: urnai/base/savable.py (hunk 2 above)
Comment (alvarofpp) [reply urnai/base/savable.py]: Yes.

**Comment by alvarofpp** [comment]
> For testing, i woud suggest first creating a class, that inherits from Savable, with some attributes. Then instantiate it, giving random values to the attributes, after that save it and load it to see if it still has the same values as before. I belive all methods of the class are used in this procedure, and that should reveal if any of them are not functional.

The Savable class is not abstract, so you should use it directly in the tests. As these are unit tests, you should ideally create tests for each function separately.

**Comment by alvarofpp** [comment]
> Should i make unit tests?

Yes, you should.

**Comment by CinquilCinquil** [comment]
What about the name Persistence for the class?

File: urnai/base/savable.py [hunk 5]
```diff
@@ -155,18 +132,18 @@ def get_pickleable_attributes(self):
...
-            self.pickle_black_list = []
+        if not hasattr(self, 'attr_block_list') or self.attr_block_list is None:
+            self.attr_block_list = []
 
         full_attr_list = [a for a in dir(self) if not a.startswith('__')
                           and not callable(getattr(self, a))
-                          and a not in self.pickle_black_list
-                          and a != 'pickle_black_list']
+                          and a not in self.attr_block_list
+                          and a != 'attr_block_list']
```
Comment (alvarofpp) [inline urnai/base/savable.py]: Rename `a` to something more explicit.

File: urnai/base/savable.py [hunk 6]
```diff
@@ -155,18 +132,18 @@ def get_pickleable_attributes(self):
         """
         This method returns a list of
         pickeable attributes. If you wish
-        to blacklist one particular
+        to blocklist one particular
         pickleable attribute, put it
-        in self.pickle_black_list as a
+        in self.attr_block_list as a
         string.
```
Comment (alvarofpp) [inline urnai/base/savable.py]: You can add more characters to the lines.

File: urnai/base/savable.py [hunk 7]
```diff
@@ -218,8 +195,8 @@ def get_pickleable_dict(self):
 
     def restore_pickleable_attributes(self, dict_to_restore):
```
Comment (alvarofpp) [inline urnai/base/savable.py]: Rename the function.

File: urnai/base/savable.py [hunk 8]
```diff
@@ -218,8 +195,8 @@ def get_pickleable_dict(self):
...
         for attr in dict_to_restore:
-            if hasattr(self, 'pickle_black_list'):
-                if attr not in self.pickle_black_list and attr != 'pickle_black_list':
+            if hasattr(self, 'attr_block_list'):
```
Comment (alvarofpp) [inline urnai/base/savable.py]: `attr_block_list` is started in `__init__`, so I can't imagine a case where it isn't in `self`.

File: urnai/base/savable.py [hunk 9]
```diff
@@ -218,8 +195,8 @@ def get_pickleable_dict(self):
 
     def restore_pickleable_attributes(self, dict_to_restore):
         for attr in dict_to_restore:
```
Comment (alvarofpp) [inline urnai/base/savable.py]:
```suggestion
for key in dict_to_restore:
```

File: urnai/base/savable.py [hunk 10]
```diff
@@ -218,8 +195,8 @@ def get_pickleable_dict(self):
...
-            if hasattr(self, 'pickle_black_list'):
-                if attr not in self.pickle_black_list and attr != 'pickle_black_list':
+            if hasattr(self, 'attr_block_list'):
+                if attr not in self.attr_block_list and attr != 'attr_block_list':
                     setattr(self, attr, dict_to_restore[attr])
             else:
                 setattr(self, attr, dict_to_restore[attr])
```
Comment (alvarofpp) [inline urnai/base/savable.py]: Avoid `setattr`.

File: urnai/base/savable.py [hunk 11]
```diff
@@ -218,8 +195,8 @@ def get_pickleable_dict(self):
 
     def restore_pickleable_attributes(self, dict_to_restore):
         for attr in dict_to_restore:
-            if hasattr(self, 'pickle_black_list'):
-                if attr not in self.pickle_black_list and attr != 'pickle_black_list':
+            if hasattr(self, 'attr_block_list'):
+                if attr not in self.attr_block_list and attr != 'attr_block_list':
                     setattr(self, attr, dict_to_restore[attr])
```
Comment (alvarofpp) [inline urnai/base/savable.py]: Avoid `setattr`.

File: urnai/base/savable.py [hunk 12]
```diff
@@ -155,18 +132,18 @@ def get_pickleable_attributes(self):
...
 
         full_attr_list = [a for a in dir(self) if not a.startswith('__')
                           and not callable(getattr(self, a))
-                          and a not in self.pickle_black_list
-                          and a != 'pickle_black_list']
+                          and a not in self.attr_block_list
+                          and a != 'attr_block_list']
         pickleable_list = []
 
         for attr in full_attr_list:
```
Comment (alvarofpp) [inline urnai/base/savable.py]:
```suggestion
for key in full_attr_list:
```

**Review by alvarofpp — CHANGES_REQUESTED** [review]

**Comment by alvarofpp** [comment]
> What about the name Persistence for the class?

It's okay for now.

File: urnai/base/savable.py (hunk 10 above)
Comment (CinquilCinquil) [reply urnai/base/savable.py]: Any suggestions on how? i've found this [page](https://snyk.io/pt-BR/blog/the-dangers-of-setattr-avoiding-mass-assignment/) and it shows mutiple ways, can i choose any of them?

**Comment by CinquilCinquil** [comment]
Btw, my lint checks keep failing because of the description of an old commit (I put too many characters in a line apparently). Is there any way to fix that?

File: urnai/base/savable.py (hunk 10 above)
Comment (alvarofpp) [reply urnai/base/savable.py]: You can use whichever approach you prefer. If there are only a few attributes, I recommend making it as explicit as possible. If it's a lot of attributes, I think it's better to validate the dictionary schema and then copy it.

**Comment by alvarofpp** [comment]
> Btw, my lint checks keep failing because of the description of an old commit (I put too many characters in a line apparently). Is there any way to fix that?

You can use interactive rebase (`git rebase -i`) for that.

File: urnai/base/savable.py [hunk 13]
```diff
@@ -0,0 +1,196 @@
...
+    This interface represents the concept of a class that can be saved to disk.
+    The heir class should define a constant or attribute as a default filename
+	to save on disk.
+    """
+
+    def __init__(self, threaded_saving=False):
+        self.threaded_saving = threaded_saving
+        self.attr_block_list = []
+
+    def get_default_save_stamp(self):
```
Comment (alvarofpp) [inline urnai/base/savable.py]: Most likely, this function will only be used by the class and not by the developer. I would add the prefix `_`.

File: urnai/base/savable.py [hunk 14]
```diff
@@ -0,0 +1,196 @@
...
+        verbosity_lvl=2)
+        '''
+
+    def threaded_pickle_save(self, savepath):
+        """
+        This method saves pickleable
+        stuff in a separate thread
+        """
+        #rp.report('THREADED Saving {} object...'
+		#.format(self.__class__.__name__), verbosity_lvl=1)
```
Comment (alvarofpp) [inline urnai/base/savable.py]: Remove comment.

File: urnai/base/savable.py [hunk 15]
```diff
@@ -0,0 +1,196 @@
...
+            self.threaded_pickle_save(savepath)
+
+        self.save_extra(savepath)
+        
+        #end_time = time.time()
+        '''
+        rp.report('It took {} seconds to save {} object!'
+		.format(end_time - start_time, self.__class__.__name__),
+        verbosity_lvl=2)
+        '''
```
Comment (alvarofpp) [inline urnai/base/savable.py]: Remove comment.

File: urnai/base/savable.py [hunk 16]
```diff
@@ -0,0 +1,196 @@
...
+        return persist_path + os.path.sep + self.get_default_save_stamp()
+
+    def save(self, savepath):
+        """
+        This method saves pickle objects
+        and extra stuff needed
+        """
+        #rp.report('Saving {} object...'.format
+		#(self.__class__.__name__), verbosity_lvl=1)
+        #start_time = time.time()
```
Comment (alvarofpp) [inline urnai/base/savable.py]: Remove comment.

File: urnai/base/savable.py [hunk 17]
```diff
@@ -0,0 +1,196 @@
...
+
+        Then all unpickleable attributes are set to None
+        and the object is pickled.
+
+        Finally the nulled attributes are
+        restored.
+        """
+        path = self.get_full_persistance_path(persist_path)
+		
+		# creating necessary directories for the file
```
Comment (alvarofpp) [inline urnai/base/savable.py]: Remove comment.

File: urnai/base/savable.py [hunk 18]
```diff
@@ -0,0 +1,196 @@
...
+
+    def get_default_save_stamp(self):
+        """
+        This method returns the default
+        file name that should be used while
+        persisting the object.
+        """
+        return self.__class__.__name__ + '_'
+
+    def save_pickle(self, persist_path):
```
Comment (alvarofpp) [inline urnai/base/savable.py]: It should be a generalized method and then a class for the pickle format that would contain the code for that specific format.

File: urnai/base/savable.py [hunk 19]
```diff
@@ -0,0 +1,196 @@
...
+            pickle.dump(self.get_pickleable_dict(), pickle_out)
+
+    def save_extra(self, persist_path):
+        """
+        This method should be implemented when
+        some extra persistence is to be saved.
+        """
+        pass
+
+    def load_pickle(self, persist_path):
```
Comment (alvarofpp) [inline urnai/base/savable.py]: It should be a generalized method and then a class for the pickle format that would contain the code for that specific format.

File: urnai/base/savable.py [hunk 20]
```diff
@@ -0,0 +1,196 @@
...
+            
+            except ValueError as ve:
+                if 'ctypes objects' in str(ve):
+                    continue
+                else:
+                    raise
+
+        return pickleable_list
+
+    def get_pickleable_dict(self):
```
Comment (alvarofpp) [inline urnai/base/savable.py]: It should be a generalized method and then a class for the pickle format that would contain the code for that specific format.

File: urnai/base/savable.py [hunk 21]
```diff
@@ -0,0 +1,196 @@
...
+
+    def load(self, loadpath):
+        """
+        This method loads pickle objects
+        and extra stuff needed
+        """
+        self.load_pickle(loadpath)
+        self.load_extra(loadpath)
+
+    def get_pickleable_attributes(self):
```
Comment (alvarofpp) [inline urnai/base/savable.py]: It should be a generalized method and then a class for the pickle format that would contain the code for that specific format.

File: urnai/base/savable.py [hunk 22]
```diff
@@ -0,0 +1,196 @@
...
+        self.save_extra(savepath)
+        
+        #end_time = time.time()
+        '''
+        rp.report('It took {} seconds to save {} object!'
+		.format(end_time - start_time, self.__class__.__name__),
+        verbosity_lvl=2)
+        '''
+
+    def threaded_pickle_save(self, savepath):
```
Comment (alvarofpp) [inline urnai/base/savable.py]: It should be a generalized method and then a class for the pickle format that would contain the code for that specific format.

File: urnai/base/savable.py [hunk 23]
```diff
@@ -0,0 +1,196 @@
...
+        some extra persistence is to be saved.
+        """
+        pass
+
+    def load_pickle(self, persist_path):
+        """
+        This method loads a list instance
+        saved by pickle.
+        """
+        # Check if pickle file exists
```
Comment (alvarofpp) [inline urnai/base/savable.py]: Remove comment.

File: urnai/base/savable.py [hunk 24]
```diff
@@ -0,0 +1,196 @@
...
+        pickle_path = self.get_full_persistance_path(persist_path)
+        exists_pickle = os.path.isfile(pickle_path)
+        
+        # If yes, load it
```
Comment (alvarofpp) [inline urnai/base/savable.py]: Remove comment.

File: urnai/base/savable.py [hunk 25]
```diff
@@ -0,0 +1,196 @@
...
+            with open(pickle_path, 'rb') as pickle_in:
+                pickle_dict = pickle.load(pickle_in)
+                self.restore_attributes(pickle_dict)
+                '''
+                rp.report(
+                    '**************************************** \n Loaded pickle for '
+                    + self.get_default_save_stamp()
+                    + ' loaded. \n****************************************',
+                2)
+				'''
```
Comment (alvarofpp) [inline urnai/base/savable.py]: Remove comment.

File: urnai/base/savable.py [hunk 26]
```diff
@@ -0,0 +1,196 @@
...
+
+        for attr in self.get_pickleable_attributes():
+            pickleable_attr_dict[attr] = getattr(self, attr)
+
+        return pickleable_attr_dict
+
+    def restore_attributes(self, dict_to_restore):
+        for key in dict_to_restore:
+            if key not in self.attr_block_list and key != 'attr_block_list':
+                setattr(self, key, dict_to_restore[key])
```
Comment (alvarofpp) [inline urnai/base/savable.py]: Avoid using `setattr`.

File: urnai/base/savable.py [hunk 27]
```diff
@@ -0,0 +1,196 @@
...
+            except AttributeError as ae:
+                if ("Can't pickle" in str(ae) or 
+                "object has no attribute '__getstate__'"):
+                    continue
+            
+            except ValueError as ve:
+                if 'ctypes objects' in str(ve):
+                    continue
+                else:
+                    raise
```
Comment (alvarofpp) [inline urnai/base/savable.py]: Remove `else`, change the logic if necessary.

File: urnai/base/savable.py [hunk 28]
```diff
@@ -0,0 +1,196 @@
...
+            
+            except NotImplementedError as nie:
+                if (str(nie) == 
+                'numpy() is only available when eager execution is enabled.'):
+                    continue
+                else:
+                    raise
```
Comment (alvarofpp) [inline urnai/base/savable.py]: Remove `else`, change the logic if necessary.

File: urnai/base/savable.py [hunk 29]
```diff
@@ -0,0 +1,196 @@
...
+            except pickle.PicklingError:
+                continue
+            
+            except TypeError as te:
+                if "can't pickle" in str(te):
+                    continue
+                if 'cannot pickle' in str(te):
+                    continue
+                else:
+                    raise
```
Comment (alvarofpp) [inline urnai/base/savable.py]: Remove `else`, change the logic if necessary.

File: urnai/base/savable.py [hunk 30]
```diff
@@ -0,0 +1,196 @@
...
+        and extra stuff needed
+        """
+        #rp.report('Saving {} object...'.format
+		#(self.__class__.__name__), verbosity_lvl=1)
+        #start_time = time.time()
+        
+        if not self.threaded_saving:
+            self.save_pickle(savepath)
+        else:
+            self.threaded_pickle_save(savepath)
```
Comment (alvarofpp) [inline urnai/base/savable.py]: Remove `else`, change the logic if necessary.

File: urnai/base/savable.py [hunk 31]
```diff
@@ -0,0 +1,196 @@
...
+
+    def threaded_pickle_save(self, savepath):
+        """
+        This method saves pickleable
+        stuff in a separate thread
+        """
+        #rp.report('THREADED Saving {} object...'
+		#.format(self.__class__.__name__), verbosity_lvl=1)
+		
+        p = Process(target=self.save_pickle, args=(savepath,))
```
Comment (alvarofpp) [inline urnai/base/savable.py]: Change the variable name to something more explicit.

**Review by alvarofpp — CHANGES_REQUESTED** [review]

File: urnai/base/savable.py (hunk 30 above)
Comment (CinquilCinquil) [reply urnai/base/savable.py]: Why?

Comment (alvarofpp) [reply urnai/base/savable.py]: This was explained at the last meeting.

File: urnai/base/persistence.py [hunk 32]
```diff
@@ -0,0 +1,83 @@
...
+    def load(self, loadpath):
+        self._load(loadpath)
+
+    @abstractmethod
+    def _load(self, savepath):
+        """
+        This method handles logic related
+        to the child class
+        """
+        ...
```
Comment (alvarofpp) [inline urnai/base/persistence.py]: Instead, you can simply set `load` to `abstractmethod`.

```suggestion
    @abstractmethod
    def load(self, loadpath):
        """
        This method handles logic related
        to the child class
        """
        ...
```

File: urnai/base/persistence.py [hunk 33]
```diff
@@ -0,0 +1,83 @@
...
+        new_process = Process(target=self._save, args=(savepath,))
+        new_process.start()
+
+    @abstractmethod
+    def _save(self, savepath):
+        """
+        This method handles logic related
+        to the child class
+        """
+        ...
```
Comment (alvarofpp) [inline urnai/base/persistence.py]: To avoid confusion with the `save` method, I would rename this to something like `simple_save`.

**Review by alvarofpp — CHANGES_REQUESTED** [review]

**Comment by alvarofpp** [comment]
Unit tests are missing.

**Review by alvarofpp — CHANGES_REQUESTED** [review]
The tests for `PersistencePickle` are missing.

**Comment by CinquilCinquil** [comment]
In the `_simple_save` and `load` tests a file is created, but i wish to delete it after the test is done. I tried `os.remove` but it needs extra permission to work. Any suggestions?

**Comment by alvarofpp** [comment]
@CinquilCinquil I recommend mocking the functions that create and read the file, that way the file won't be created and won't need to be removed later.

**Comment by CinquilCinquil** [comment]
Oh, but the whole purpose of `test_simple_save`, for example, is to check if a file has been created. I'm not familiar with mocking, is  there a way of faking file creation?

**Comment by alvarofpp** [comment]
Creating files is something that can result storage costs, and you have to make sure that each test involving `_simple_save` generates a different file so that you don't have the problem of trying to save to the same file or the file not existing because of another test. In short, it's something that can cause more problems than it solves. To test this, I believe that flow/process tests are more suitable than unit tests.

What you can do is mock the function that saves or reads the file and change its execution, in which case you should apply the mock to the `return_value` of the function.

File: urnai/base/persistence_pickle.py [hunk 34]
```diff
@@ -0,0 +1,111 @@
...
+        saved using pickle, the ones which are not
+        are backuped.
+
+        Then all unpickleable attributes are set to None
+        and the object is pickled.
+
+        Finally the nulled attributes are
+        restored.
+        """
+        path = super().get_full_persistance_path(persist_path)
```
Comment (alvarofpp) [inline urnai/base/persistence_pickle.py]:
```suggestion
        path = self.get_full_persistance_path(persist_path)
```

File: urnai/base/persistence_pickle.py [hunk 35]
```diff
@@ -0,0 +1,111 @@
...
+		
+        with open(path, 'wb') as pickle_out:
+            pickle.dump(self._get_dict(), pickle_out)
+
+    def load(self, persist_path):
+        """
+        This method loads a list instance
+        saved by pickle.
+        """
+        pickle_path = super().get_full_persistance_path(persist_path)
```
Comment (alvarofpp) [inline urnai/base/persistence_pickle.py]:
```suggestion
        pickle_path = self.get_full_persistance_path(persist_path)
```

File: urnai/base/persistence_pickle.py [hunk 36]
```diff
@@ -0,0 +1,111 @@
...
+        """
+        if not hasattr(self, 'attr_block_list') or self.attr_block_list is None:
+            self.attr_block_list = []
+
+        full_attr_list = [attr for attr in dir(self) if not attr.startswith('__')
+                          and not callable(getattr(self, attr))
+                          and attr not in self.attr_block_list
+                          and attr != 'attr_block_list'
+                          and attr != 'processes'
+                          and 'abc' not in attr]
```
Comment (alvarofpp) [inline urnai/base/persistence_pickle.py]:
```suggestion
        if not hasattr(self, 'attr_block_list') or self.attr_block_list is None:
            self.attr_block_list = []

        attr_block_list = self.attr_block_list + ['attr_block_list', 'processes']

        full_attr_list = [attr for attr in dir(self) if not attr.startswith('__')
                          and not callable(getattr(self, attr))
                          and attr not in attr_block_list
                          and 'abc' not in attr]
```

File: urnai/base/persistence_pickle.py [hunk 37]
```diff
@@ -0,0 +1,111 @@
...
+                continue
+            
+            except AttributeError as attr_error:
+                if ("Can't pickle" in str(attr_error) or 
+                "object has no attribute '__getstate__'"):
+                    continue
+            
+            except ValueError as value_error:
+                if 'ctypes objects' not in str(value_error):
+                    raise
```
Comment (alvarofpp) [inline urnai/base/persistence_pickle.py]: Raise what?

File: urnai/base/persistence_pickle.py [hunk 38]
```diff
@@ -0,0 +1,111 @@
...
+                
+                pickleable_list.append(key)
+                
+            except pickle.PicklingError:
+                continue
+            
+            except TypeError as type_error:
+                if ("can't pickle" not in str(type_error) or
+                 'cannot pickle' not in str(type_error)):
+                    raise
```
Comment (alvarofpp) [inline urnai/base/persistence_pickle.py]: Raise what?

File: urnai/base/persistence_pickle.py [hunk 39]
```diff
@@ -0,0 +1,111 @@
...
+                continue
+            
+            except NotImplementedError as notimpl_error:
+                if (str(notimpl_error) != 
+                'numpy() is only available when eager execution is enabled.'):
+                    raise
```
Comment (alvarofpp) [inline urnai/base/persistence_pickle.py]: Raise what?

**Review by alvarofpp — CHANGES_REQUESTED** [review]

**Review by alvarofpp — APPROVED** [review]

**Pull request merged**